java-checks/src/main/java/org/sonar/java/checks/CollectorsToListCheck.java

```diff
@@ -98,42 +107,82 @@ protected void onMethodInvocationFound(MethodInvocationTree mit) {
   }
 
   private static boolean isReturnedListUnmodified(MethodInvocationTree mit) {
-    Symbol assignedVariable = findAssignedVariable(mit.parent());
-    return assignedVariable == null || assignedVariable.usages().stream().noneMatch(CollectorsToListCheck::isListBeingModified);
-  }
-
-  private static boolean isInvariantTypeArgument(MethodInvocationTree collectMethodInvocation) {
-    Type streamType = collectMethodInvocation.methodSymbol().owner().type();
-    if (streamType.isRawType()) {
+    var graylist = new HashSet<Symbol>();
+    if (isPossibleMutatorExpression(mit, graylist)) {
+      return false;
+    }
+    if (graylist.isEmpty()) {
       return true;
     }
-    Type collectArgType = collectMethodInvocation.symbolType().typeArguments().get(0);
-    Type streamArgType = streamType.typeArguments().get(0);
-    return collectArgType.is(streamArgType.fullyQualifiedName());
-  }
 
-  private void reportIssue(MethodInvocationTree collector, boolean mutable) {
-    String message = String.format(MESSAGE, mutable ? "toList" : "toUnmodifiableList");
-    reportIssue(collector, message);
+    var whiteList = new HashSet<Symbol>();
+    do {
+      final var nextGrayList = new HashSet<Symbol>();
+      if (graylist.stream().anyMatch(symbol -> isPossibleMutatorSymbol(symbol, nextGrayList))) {
+        return false;
+      }
+      whiteList.addAll(graylist);
+      nextGrayList.removeAll(whiteList);
+      graylist = nextGrayList;
+    } while (!graylist.isEmpty());
+    return true;
   }
 
-  private static boolean isListBeingModified(Tree list) {
-    return isListTargetOfCallAndBeingModified(list) || isListArgumentToCollectionsMutatorMethod(list);
+  private static boolean isPossibleMutatorSymbol(Symbol symbol, Set<Symbol> graylist) {
+    return symbol.usages().stream().anyMatch(it -> isPossibleMutatorExpression(it, graylist));
   }
 
-  private static boolean isListArgumentToCollectionsMutatorMethod(Tree list) {
-    if (list.parent().is(Tree.Kind.ARGUMENTS)) {
-      Tree potentialMethodInvocation = list.parent().parent();
-      if (potentialMethodInvocation.is(Tree.Kind.METHOD_INVOCATION)) {
-        MethodInvocationTree mit = (MethodInvocationTree) potentialMethodInvocation;
-        return COLLECTIONS_MUTATOR_METHODS.matches(mit);
+  private static boolean isPossibleMutatorExpression(ExpressionTree tree, Set<Symbol> graylist) {
+    var parent = tree.parent();
+    return switch (parent.kind()) {
+      case LAMBDA_EXPRESSION ->
+        // Don't perform deeper data flow analysis; assume receiver of lambda result might perform mutations
```

Comment:
I would not mention data flow analysis here... we are not looking at any flow here, just at the parent tree...